File: framework/languages/java/e2e/RegistryMessaging/Client/src/main/java/systems/zlink/e2e/registrymessaging/client/Scenarios/RmC4TimeoutIsolationScenario.java
package systems.zlink.e2e.registrymessaging.client.Scenarios;

import systems.zlink.e2e.registrymessaging.client.Support.ScenarioAssert;
import systems.zlink.e2e.registrymessaging.client.Support.ScenarioWait;
import systems.zlink.e2e.registrymessaging.shared.Contracts;
import systems.zlink.httpclient.ZLinkHttpClient;

public final class RmC4TimeoutIsolationScenario {
    private RmC4TimeoutIsolationScenario() {
    }

    public static void run(
        ZLinkHttpClient discoveryConsumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB) {
        Contracts.RequestFailureRes timeout = discoveryConsumer.post("/profile/slow-request")
            .body(new Contracts.ProfileReq("slow"))
            .submit(Contracts.RequestFailureRes.class).toCompletableFuture().join().body();
        ScenarioAssert.that(timeout.failed(), "RM-C4 expected slow request failure");
        ScenarioAssert.that("TimeoutException".equals(timeout.errorKind()),
            "RM-C4 expected public TimeoutException, got " + timeout.errorKind());
        Contracts.ProfileRes after = discoveryConsumer.post("/profile/request")
            .body(new Contracts.ProfileReq("after"))
            .submit(Contracts.ProfileRes.class).toCompletableFuture().join().body();
        ScenarioAssert.that("profile:after".equals(after.value()), "RM-C4 post-timeout request failed");
        ScenarioWait.sleep(1100);
        Contracts.ProfileRes later = discoveryConsumer.post("/profile/request")
            .body(new Contracts.ProfileReq("later"))
            .submit(Contracts.ProfileRes.class).toCompletableFuture().join().body();
        ScenarioAssert.that("profile:later".equals(later.value()), "RM-C4 later request failed");
        String[] slowEvidence = ScenarioAssert.waitAnyEvidence(providerA, providerB, "value=slow");
        ScenarioAssert.that(java.util.Arrays.stream(slowEvidence)
                .anyMatch(line -> line.contains("ProfileReq") && line.contains("value=slow")),
            "RM-C4 slow handler completion evidence missing");
        System.out.println("scenario RM-C4 passed");
    }
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/client/Scenarios/RmC4TimeoutIsolationScenario.kt
package Scenarios

import systems.zlink.e2e.kotlin.registrymessaging.client.Scenarios
import systems.zlink.e2e.kotlin.registrymessaging.client.Support
import systems.zlink.e2e.kotlin.registrymessaging.client.Support.HttpJson
import systems.zlink.e2e.kotlin.registrymessaging.client.Support.ScenarioAssert
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.RequestFailureRes

object RmC4TimeoutIsolationScenario {
    fun run(discoveryConsumer: HttpJson, providerA: HttpJson, providerB: HttpJson) {
        val timeout = discoveryConsumer.post<RequestFailureRes>("/profile/slow-request", ProfileReq("slow"))
        ScenarioAssert.that(timeout.failed, "RM-C4 expected the slow request to time out.")

        val immediate = discoveryConsumer.post<ProfileRes>("/profile/request", ProfileReq("rm-c4-after-timeout"))
        ScenarioAssert.that(immediate.value == "profile:rm-c4-after-timeout", "RM-C4 follow-up reply mismatch.")
        Thread.sleep(1200)
        val later = discoveryConsumer.post<ProfileRes>("/profile/request", ProfileReq("rm-c4-later"))
        ScenarioAssert.that(later.value == "profile:rm-c4-later", "RM-C4 later reply mismatch.")

        val evidence = providerA.get<List<String>>("/evidence") + providerB.get<List<String>>("/evidence")
        ScenarioAssert.that(
            evidence.any { it.contains("rm-c4-after-timeout") } && evidence.any { it.contains("rm-c4-later") },
            "RM-C4 follow-up evidence missing.",
        )
        println("scenario RM-C4 passed")
    }
}
